indico/modules/events/models/events.py

```diff
@@ -669,6 +669,20 @@ def can_lock(self, user):
         """Check whether the user can lock/unlock the event"""
         return user and (user.is_admin or user == self.creator or self.category.can_manage(user))
 
+    def can_display(self):
+        """Check whether the user can display the event in the category.
+        To be replaced with new permission mechanism: can('display', self, session.user)."""
+        return self.visibility is None or self.can_manage(session.user)
+
+    @staticmethod
+    def get_hidden_events(category_id):
+        """Get all hidden events within the given category."""
+        category = Category.get(int(category_id), is_deleted=False)
+        return (Event.query.with_parent(category)
+                .filter(Event.category_id == category_id, Event.visibility == 0)
+                .order_by(Event.start_dt.desc(), Event.id.desc())
+                .all())
+
     def get_relative_event_ids(self):
```

Comment:
I think you should exclude hidden events in this function as well (since you cannot easily check can_display there). But for this type of navigation (the next/prev/first/last buttons on top of an event display page) excluding those events is probably fine anyway.

Reply:
Not clear... so can I keep it as is?

Reply:
.where((Event.category_id == self.category_id) & ~Event.is_deleted)

You need to add ` & (Event.visibility != 0)` here (and test it, I'm not sure if it works correctly due to how NULL behaves, you may need something more explicit such as `(Event.visibility.is_(None) | Event.visibility != 0))`

Reply:
Wait... here I need to retrieve **only** the **invisible** ones... so visibility=0.
As you proposed:
hidden_events = Event.query.with_parent(self).filter_by(visibility=0)

Reply:
Check which function this comment is on ;)

Reply:
Ahhhh ok, I got it now! :)